Add version lock

EmmanuelVella opened the pull request:
This PR implements object locking.

When loading an object for edit, its original version number is stored in the form. When submitting the form, this number is then passed to the ``Admin::lock()`` method. If a ``LockException`` is thrown, a flash message will inform the user that the object has been edited by someone.

If the ``ModelManager`` does not implements the LockInterface, the object will be updated normally.

Refs #2655

OskarStark commented:
thank you for your work!

this is a great and often needed feature :+1:

rande commented on Controller/CRUDController.php:
```diff
@@ -411,8 +415,12 @@ public function editAction($id = null, Request $request = null)
             // persist if the form was valid and if in preview mode the preview was approved
             if ($isFormValid && (!$this->isInPreviewMode($request) || $this->isPreviewApproved($request))) {
                 try {
+                    $this->admin->lock($object, $session->get($versionKey));
```
I think the version check can be done in an extension.

mvhirsch commented:
:+1: awesome feature, please add it soon!

rande commented on Admin/AdminInterface.php:
```diff
@@ -656,6 +657,21 @@ public function delete($object);
      *
+     * @return int|null
+     */
+    public function getVersion($object);
```
This should be part of the model manager interface.

rande replied:
Actually there is no need for this as the ``LockInterface`` interface already exist in this PR.

EmmanuelVella replied:
@rande If I remove these methods from the AdminInterface, that means that we need to call them directly on the ``ModelManager`` from the ``CRUDController::editAction``. Is it ok ?

EmmanuelVella replied:
Ok So I just make the Admin class to implement the LockInterface, right ?

rande replied:
no, only the ModelManager.

rande replied:
@EmmanuelVella you can only have the catch in the ``CRUDController`` to set the correct error message. However the check can be done inside an extension. see: the ``AdminExtension`` class and the related documentation: https://sonata-project.org/bundles/admin/2-3/doc/reference/extensions.html

rande commented on Admin/AdminInterface.php:
```diff
@@ -656,6 +657,21 @@ public function delete($object);
+     *
+     * @throws LockException
+     */
+    public function lock($object, $expectedVersion);
```
This should be part of the model manager interface.

Soullivaneuh commented on Admin/AdminInterface.php:
```diff
@@ -656,6 +657,21 @@ public function delete($object);
+    public function lock($object, $expectedVersion);
+
+    /**
+     * @param mixed $object
```
`@param object` instead?

EmmanuelVella replied:
Right, I fix it

Soullivaneuh replied:
You have many of them to fix. :+1:

EmmanuelVella replied:
Fixed

Soullivaneuh replied:
Great. Please also consider https://github.com/sonata-project/SonataAdminBundle/pull/3170#issuecomment-129924006.

EmmanuelVella replied:
Sure, I will do it when all is ok

Soullivaneuh commented:
If I understand well, the lock logic will be automatically functional without configuration.

I like this system, but I don't like the way to force it.

I think we should let the choice with an option like `lock_protection`. This option should be `false` by default to be BC.

OskarStark commented on Admin/Admin.php:
```diff
@@ -697,6 +698,30 @@ public function delete($object)
     /**
      * {@inheritdoc}
      */
+    public function getVersion($object)
```
IMO i would rename this method to something like ```getLockVersion()``` otherwise the could be some understanding problems with history/versioning

Soullivaneuh replied:
> otherwise the could be some understanding problems with history/versioning

:+1:

EmmanuelVella replied:
Absolutely, fixed

rande commented on Resources/translations/SonataAdminBundle.en.xliff:
```diff
@@ -162,6 +162,10 @@
             </trans-unit>
+            <trans-unit id='flash_lock_error'>
+                <source>flash_lock_error</source>
+                <target>Another user has modified item "%name%". Please reload the page and apply the changes again.</target>
```
I think the message should be: Another user has modified item "%name%". Please <a href="...">click here</a> to reload the page and apply the changes again. Maybe we can use some font awesome icon too.

rande replied:
<script>alert('XXX'); </script>

EmmanuelVella replied:
Updated

EmmanuelVella commented on Controller/CRUDController.php:
```diff
@@ -456,6 +473,8 @@ public function editAction($id = null, Request $request = null)
                 $this->admin->getShow();
             }
+        } else {
+            $session->set($versionKey, $this->admin->getVersion($object));
```
@rande If I move the locking logic in an extension, how can I achieve this ? I need to do this only when the object is loaded for the first time and the form is not submited yet.

EmmanuelVella replied:
@rande Please tell me if you see a way to detect if the form has been submitted from an ``AdminExtension``.

The only way I see is to add a new method in the ``AdminExtension`` class, and call it somewhere after ``$form->handleRequest($request)``.

EmmanuelVella replied:
@rande The problem is that at this moment, the object version number will be hydrated from the database. We need to remember the version number when loading the form for the first time.

Let me try to add the version number in the form, which will be more complex to achieve but maybe more decoupled ?

rande replied:
Having this version number in the form will avoid some possible weird behavior with session.

rande commented:
@EmmanuelVella you can use the ``preUpdate`` function to get the object. It will be hydrated with the version number.

Soullivaneuh commented on Exception/LockException.php:
```diff
@@ -0,0 +1,21 @@
+namespace Sonata\AdminBundle\Exception;
+
+/**
+ * Class LockException.
```
Useless class comment.

Soullivaneuh commented on Model/LockInterface.php:
```diff
@@ -0,0 +1,37 @@
+use Sonata\AdminBundle\Exception\LockException;
+
+/**
+ * LockInterface.
```
Useless class comment.

rande commented:
@EmmanuelVella you also need to check the CS (make cs)

EmmanuelVella commented:
I have updated my PR to use a form field instead of the session to store the initial version. This is cleaner and more decoupled. Now, all is done in the ``LockExtension``.

@Soullivaneuh I also added the ``lock_protection`` option.

rande commented on Admin/Extension/LockExtension.php:
```diff
@@ -0,0 +1,111 @@
+
+    public function __construct()
+    {
+        $this->lockVersions = new \SplObjectStorage();
```
Why do we need to keep a global state ?

EmmanuelVella replied:
As it's a service, it was to avoid potential bugs. But I'm not sure it's useful. To you prefer a simple $version variable ?

rande replied:
the question is more about why do you need to keep object across functions ?

Soullivaneuh commented:
Thanks for the option.

Maybe a little cookbook page about this feature could be helpful. ;-)

rande commented on Admin/Extension/LockExtension.php:
```diff
@@ -0,0 +1,106 @@
+ */
+class LockExtension extends AdminExtension
+{
+    protected $version;
```
No, I don't say you need to rename the variable. I just need to understand why you need to keep a state in the service.

EmmanuelVella replied:
It's not just a rename, it's just an int now.

I need to keep the original version number because when the form is submitted, the object will be hydrated from the database data, and the version number may have increased (if someone else has edited the object). So I can't rely on this updated version number to lock the object.

EmmanuelVella replied:
When calling ``LockExtension::preUpdate``, I need to get the version number which was in the form field, that's why I need to store it during the ``PRE_SUBMIT`` event.

rande replied:
Well, if you post the version field, the posted version will be used when the POST data will be mapped to the object. So there is no need to store the version in the extension.

EmmanuelVella replied:
We can't map the ``_lock_version`` field to the object, because doctrine reads the version number directly from the object (not from the database) and then compare it to the version number we use when calling ``lock()``: https://github.com/doctrine/doctrine2/blob/v2.5.0/lib/Doctrine/ORM/UnitOfWork.php#L2326

EmmanuelVella commented:
@Soullivaneuh I have added a small cookbook entry : https://github.com/sonata-project/SonataAdminBundle/pull/3170/files#diff-8d214942c495e756c0181bf3658710b9R1

Tell me if you want to add/modify something !

OskarStark commented on Resources/doc/cookbook/recipe_lock_protection.rst:
```diff
@@ -0,0 +1,31 @@
+                lock_protection: true
+
+.. note::
+    Lock protection must be supported by the corresponding model manager.
```
how to achieve this?
or is it supported by the default ModelManager?

rande replied:
@OskarStark see https://github.com/sonata-project/SonataDoctrineORMAdminBundle/pull/478/files

OskarStark replied:
yes, i know, but the note then is confusing, so i would write it like this:
```Lock protection must be supported by the corresponding model manager, which is already done if you use the default ModelManager```

and then we should write down that this feature is just for the DoctrineORM, isn't it?

EmmanuelVella replied:
@OskarStark The ``SonataDoctrineORMAdminBundle`` is not installed/required when you install the ``SonataAdminBundle`` (it's not the "default" admin doctrine bundle).

However, we may add a note saying that currently only the ``SonataDoctrineORMAdminBundle`` supports it.

rande replied:
@EmmanuelVella I think @OskarStark knows that ;) his comment is more about the completeness of the documentation.

OskarStark replied:
yes :smile: :+1:

EmmanuelVella replied:
Hehe ok, I have updated the note accordingly. Tell me if it's more clear.

rande commented on Model/LockInterface.php:
```diff
@@ -0,0 +1,35 @@
+
+    /**
+     * @param object $object
+     * @param int    $expectedVersion
```
should be mixed, you don't know the implementation

EmmanuelVella replied:
Fixed

rande commented:
We are almost there I think!

EmmanuelVella commented:
\o/

OskarStark commented on Resources/doc/cookbook/recipe_lock_protection.rst:
```diff
@@ -0,0 +1,33 @@
+.. note::
+    If the object model manager does not supports object locking,
+    the lock protection will not be triggered for the object.
+    Currently, only the `SonataDoctrineORMAdminBundle` supports it.
```
:+1: like it, thank you!

@rande or do want to link to the DOC of the ORM Bundle?

EmmanuelVella commented:
:+1: Thank you all

OskarStark commented:
Thanks for your great work @EmmanuelVella

mcwebb commented:
Am I right in thinking that for this feature to work, Optimistic Locking needs to be configured for each entity?

EmmanuelVella commented:
You need to use the ``Version` annotation in each entity you want to lock : http://doctrine-orm.readthedocs.org/projects/doctrine-orm/en/latest/reference/annotations-reference.html#version

mcwebb commented:
Thank you, this isn't clear from the cookbook. I would suggest that info is added.

OskarStark commented:
could you please provide a PR @mcwebb ?

mcwebb commented:
will do

OskarStark commented:
thank you :+1:

Soullivaneuh commented:
@EmmanuelVella Would not be interesting to add this option customizable on each admin definition in addition of the global setting?

mcwebb commented:
@OskarStark sorted. https://github.com/sonata-project/SonataAdminBundle/pull/3379
Let me know if it needs to be tweaked.

EmmanuelVella commented:
@Soullivaneuh It's possible, but what would be the benefit ?

Soullivaneuh commented:
> but what would be the benefit ?

If we want to enable lock protection only for some entities.

BTW, I think some peoples will want to use doctrine `@Version` but not version lock of sonata, for some cases.
